Atomic/bond targets prediction

shihchengli opened the pull request:
## Description
This PR adds atom- and bond-level properties prediction to the Chemprop. It can train on any number of atomic/bond properties simultaneously with multitask D-MPNN. It also supports the capability to carry out uncertainty quantification, calibration, and evaluation. Besides, the constraints of a sum to a specific atomic/bond properties for each molecule can be applied. The prediction results are saved as a `.csv` file.

The current model can only be trained on `dataset_type` of regression or classification (multiclass is not supported).

- --is_atom_bond_targets - whether this is atomic/bond properties prediction.

- --constraints_path - path to a `.csv` file to use for constraints. The constraints must be in the same order as the SMILES strings in the provided data file.

## Example / Current workflow
Workflow of training multitask model on one atomic target and two bond targets with constraints.
```
Chemprop_train \
--data_path /path/to/train.csv \
--dataset_type regression \
--is_atom_bond_targets \
--constraints_path /path/to/constraints.csv \
--adding_h
```

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

jonwzheng reviewed (commented):
Good work on the atom + bond-level targets. I ran this code using the 1K subsplit of Yanfei's QM data you provided. I only had time to take a quick look, but left some notes (&I will look more later). Apart from the comments I left here, I also noticed that during training, I saw persistent warning messages:
```
/home/gridsan/jonzheng/.local/lib/python3.8/site-packages/sklearn/utils/validation.py:63: FutureWarning: Arrays of bytes/strings is being converted to decimal numbers if dtype='numeric'. This behavior is deprecated in 0.24 and will be removed in 1.1 (renaming of 0.26). Please convert your data to numeric values explicitly instead.
```

This is coming from function calls to the loss function in chemprop's `metrics.py`, or more specifically, in `_check_reg_targets` in `sklearn.metrics._regression.py`. I think that the targets need to be sanitized to floats as they're read in before they get to this step in chemprop.

shihchengli commented on README.md:
```diff
@@ -302,6 +303,20 @@ reaction and solvent/molecule encoding. Below are the input arguments for specif
   * `--depth_solvent` Number of message passing steps for solvent/molecule.
   * `--adding_h` Whether RDKit molecules will be constructed with adding the Hs to them. Applicable to any SMILES that is not reaction.
```
Yes! That is a good point. Thanks!

cjmcgill commented on README.md:
```diff
@@ -302,6 +303,20 @@ reaction and solvent/molecule encoding. Below are the input arguments for specif
 
+### Atomic and bond properties prediction
+
+Chemprop can perform multitask constraint message passing neural networks for atomic/bond properties prediction as described in this [paper](https://chemrxiv.org/articles/preprint/Regio-Selectivity_Prediction_with_a_Machine-Learned_Reaction_Representation_and_On-the-Fly_Quantum_Mechanical_Descriptors/12907316). This model can train on any number of atomic/bond properties simultaneously. In the or ...
```
```suggestion
Chemprop can perform multitask constrained message passing neural networks for atomic/bond properties prediction as described in this [paper](https://chemrxiv.org/articles/preprint/Regio-Selectivity_Prediction_with_a_Machine-Learned_Reaction_Representation_and_On-the-Fly_Quantum_Mechanical_Descriptors/12907316). This model can train on any number of atomic/bond properties simultaneously. In the original work, a total loss was calculated as a weighted sum of every single loss, where the weights were required to be specified for the regression task. In this repository, these weights have been automatically taken into account by doing standardization of all the training targets. In order to train a model, training data containing molecules (as SMILES strings) and known atomic/bond target values are required. The input format can either be a csv or pickle file. For example:
```

cjmcgill commented on README.md:
````diff
@@ -302,6 +303,20 @@ reaction and solvent/molecule encoding. Below are the input arguments for specif
+```
+where atomic properties (e.g. hirshfeld_charges) must be a 1D list with the order same as that of atoms in the SMILES string; and bond properties (e.g. bond_length_matrix) can either be a 2D list of shape (number_of_atoms × number_of_atoms) or a 1D list with the order same as that of bonds in the SMILES string. If the input format is a pickle file, these targets can also be saved as numpy arrays.
+The atomic/bond training targets can be assigned by using `--atom_targets` and `--bond_targets`. This model allows multitask constraints applied to different atomic/bond properties by specifying the argument `--constraints_path` with a given `.csv` file. Note that the constraints must be in the same order as the SMILES strings in your data file. Also note that `.csv` file must have a header row a ...
+
````
Seems like the description in L318 is accidentally unfinished.

cjmcgill replied:
If you provide constraints for one task, do you have to provide constraints for all tasks? What happens if you have blanks in the constraints file? Also probably worth mentioning that only atom-targets can be constrained.

cjmcgill replied:
Oh, I see below that it chooses if tasks are constrained based off of the header. I don't think there is otherwise an error that triggers if there are multiple copies of a task name in a header. You might need to add that to avoid this edge case.

cjmcgill commented on chemprop/args.py:
```diff
@@ -229,6 +233,10 @@ class TrainArgs(CommonArgs):
     """
     ignore_columns: List[str] = None
     """Name of the columns to ignore when :code:`target_columns` is not provided."""
+    atom_targets: List[str] = []
+    """Name of the columns containing target atomic values."""
+    bond_targets: List[str] = []
+    """Name of the columns containing target bond values."""
```
Would these be better suited as TrainArgs? With a separate argument in PredictArgs?

cjmcgill commented on chemprop/args.py:
```diff
@@ -702,7 +760,12 @@ def process_args(self) -> None:
                     raise ValueError(f'Additional features were provided using the argument {features_argument}. The same kinds of features must be provided for the separate validation set.')
                 if self.separate_test_path is not None and test_features_path is None:
                     raise ValueError(f'Additional features were provided using the argument {features_argument}. The same kinds of features must be provided for the separate test set.')
```
You should add the `--separate_val_constraints_path` and `--separate_test_constraints_path` to the checks in L751-762.

cjmcgill commented on chemprop/args.py:
```diff
@@ -546,14 +582,37 @@ def process_args(self) -> None:
+        self.is_atom_bond_targets = False
+        if self.atom_targets != [] or self.bond_targets != []:
+            self.is_atom_bond_targets = True
+            self.target_columns = self.atom_targets + self.bond_targets
```
So we're not enabling molecule-targets alongside atom-bond?

shihchengli replied:
My first thought is to let chemprop have the capability to train on atom- and bond-level targets. The predicted atom- and bond-level targets can further be used as additional features in molecule properties prediction. However, It makes sense that people would want to train on atom-level, bond-level, and molecule targets at the same time because it might improve the performance of the model. Do you think it would be better to support this feature in current PR? It might just take a few days to finish this part.

cjmcgill replied:
Having mixed targets seems like a special case. We can leave it off for now unless you see a convenient way to implement it.

cjmcgill replied:
Though, I will say, it's possible our current project would benefit from mixed targets.

cjmcgill commented on chemprop/args.py:
```diff
@@ -546,14 +582,37 @@ def process_args(self) -> None:
+        self.is_atom_bond_targets = False
+        if self.atom_targets != [] or self.bond_targets != []:
+            self.is_atom_bond_targets = True
+            self.target_columns = self.atom_targets + self.bond_targets
```
Does this just overwrite the `--target_columns` argument? That was useful for allowing certain targets in a data file to be ignored. That would sill probably be useful for atom/bond operations.

shihchengli replied:
Because the current version cannot train atom-, bond-, and molecule-level targets at the same time, it won't overwrite the `--target_columns` when training on atom- and bond-level targets. However, if training on atom-, bond-, and molecule-level targets is expected, this code should be modified.

shihchengli replied:
I have modified this part. The `--atom-targets` and `--bond-targets` arguments have been removed. User can use `--target_columns` to specify the targets they want to train. However, if atomic or bond targets have been specified, the molecule-level targets would be ignored.

cjmcgill commented on chemprop/args.py:
```diff
@@ -546,14 +582,37 @@ def process_args(self) -> None:
+
+        # Check whether the number of input columns is one for the atomic/bond mode
+        if self.is_atom_bond_targets:
+            if len(self.smiles_columns) != 1:
```
Better to trigger this off of `self.number_of_molecules`

cjmcgill commented on chemprop/args.py:
```diff
@@ -546,14 +582,37 @@ def process_args(self) -> None:
+        # Check whether the argument of `--adding_h` is used
+        if self.is_atom_bond_targets:
+            if not self.adding_h:
+                raise ValueError('In atomic/bond properties prediction, argument `--adding_h` should be used.')
```
Should this always be the case? Couldn't you have places where people are just training off of heavy atoms?

shihchengli replied:
It is a good point. If people want to train on heavy atoms, this error message doesn't make sense. This code would be removed. Besides, I would add some descriptions in README.md to notice people the argument of `--adding_h` should be used when training on atoms including hydrogen atoms.

cjmcgill commented on chemprop/data/utils.py:
```diff
@@ -8,22 +9,33 @@
+from chemprop.rdkit import make_mol
+
+# Increase maximum size of field in the csv processing
+csv.field_size_limit(sys.maxsize)
```
Why is this coming up?

shihchengli replied:
Line 19: `make_mol` is used to create RDKit molecule object in Line 416.

Line 21-22: For atom- and bond-level training, users can use either csv or pickle as input file. When using csv as input, it might appear `_csv.Error: field larger than field limit (131072)`, which means the csv file might contain very huge fields. This makes sense because the arrays of atom- and bond-level targets might be very large. In order to solve this, the field limit is set to sys.maxsize.

cjmcgill commented on chemprop/data/utils.py:
```diff
@@ -8,22 +9,33 @@
-        header = next(csv.reader(f))
+    extension = os.path.splitext(path)[1]
+
+    if extension in ['.pkl', '.pckl', '.pickle']:
+        df = pd.read_pickle(path)
+        header = df.columns.tolist()
```
This function (maybe others too) in the data utils file could use unit tests showing that they can handle pkl inputs too.

cjmcgill commented on chemprop/args.py:
```diff
@@ -523,6 +535,30 @@ def bond_feature_scaling(self) -> bool:
+            header = chemprop.data.utils.get_header(self.constraints_path)
+            atom_constraints = [target in header for target in self.atom_targets]
+        else:
+            atom_constraints = [None] * len(self.atom_targets)
```
Why [None] and not [False]? Same at L559.

shihchengli replied:
Thanks! It should be `False`.

cjmcgill commented on chemprop/args.py:
```diff
@@ -229,6 +233,10 @@ class TrainArgs(CommonArgs):
     """
     ignore_columns: List[str] = None
     """Name of the columns to ignore when :code:`target_columns` is not provided."""
+    atom_targets: List[str] = []
+    """Name of the columns containing target atomic values."""
+    bond_targets: List[str] = []
+    """Name of the columns containing target bond values."""
```
If these two were left blank, any way we could figure out which targets belonged in which columns? Like under normal molecule-level operation you would just run `--get_task_names` and assume that they were all target columns by default. Is there a data utils function that could be addressed to make these default more easily?

shihchengli replied:
If they were left blank, this model would not do anything. I would build a data utils function to address this. Really appreciate this suggestion.

cjmcgill replied:
If you think there's a function that can determine which atom/bonds are which and it works really well, then you wouldn't even need the atom_feature_targets and bond_feature_targets arguments. You could remove them and use just target_columns to say which of those to use and which to ignore (using all by default). If choosing which ones are atom and which are bond is hard or unreliable though, then the current set up would be better.

shihchengli replied:
That is a good idea! I added a `get_mixed_task_names()` function in data/utils.py, `--atom_targets` and `--bond_targets` have been removed and `--target_columns` becomes the default option.

cjmcgill commented on README.md:
````diff
@@ -302,6 +303,20 @@ reaction and solvent/molecule encoding. Below are the input arguments for specif
+4      CC(=N)NCc1cccc(CNCc2ccncc2)c1  [-0.083162, 0.114954, -0.274544, -0.100369, 0....  ...  [[0.0, 1.5137126697008916, 2.4882198180715465,...  [[0.0, 1.0036, 0.0437, 0.0108, 0.0134, 0.0004,......
+```
+where atomic properties (e.g. hirshfeld_charges) must be a 1D list with the order same as that of atoms in the SMILES string; and bond properties (e.g. bond_length_matrix) can either be a 2D list of shape (number_of_atoms × number_of_atoms) or a 1D list with the order same as that of bonds in the SMILES string. If the input format is a pickle file, these targets can also be saved as numpy arrays.
+The atomic/bond training targets can be assigned by using `--atom_targets` and `--bond_targets`. This model allows multitask constraints applied to different atomic/bond properties by specifying the argument `--constraints_path` with a given `.csv` file. Note that the constraints must be in the same order as the SMILES strings in your data file. Also note that `.csv` file must have a header row a ...
````
```suggestion
The atomic/bond training targets can be assigned by using `--atom_targets` and `--bond_targets`. This model allows multitask constraints applied to different atomic/bond properties by specifying the argument `--constraints_path` with a given `.csv` file. Note that the constraints must be in the same order as the SMILES strings in your data file. Also note that `.csv` file must have a header row and the constraints should be comma-separated with one line per molecule. The `--adding_h` option can be used if hydrogens are included in the atoml targets and bonds to hydrogens are included in the bond targets.
```

cjmcgill commented on chemprop/data/data.py:
```diff
@@ -175,6 +188,24 @@ def number_of_molecules(self) -> int:
+        :return: The number of bonds.
+        """
+        return len(self.mol[0].GetBonds())
+
```
This is a good addition. We should have added these properties a long time ago.

shihchengli replied:
Thanks! However, it only returns the value of the first molecule (`--number_of_molecules` is 1). I should modify this to return an array containing the values for each molecule (`--number_of_molecules` > 1).

cjmcgill commented on chemprop/data/data.py:
```diff
@@ -510,7 +589,7 @@ def normalize_features(self, scaler: StandardScaler = None, replace_nan_token: i
         return scaler
 
-    def normalize_targets(self) -> StandardScaler:
+    def normalize_targets(self) -> Union[List[StandardScaler], StandardScaler]:
```
Needing to have every location for the scaler variable support both StandardScaler and List[StandardScaler] is going to be very cumbersome. It's probably taking a lot of code to support the ambiguity now, and I expect it will make development around it more confusing in the future. I am not sure that it's necessary.

A single Scaler under molecule-level targets will carry a mean and a stdev for each task. Couldn't we have a single Scaler carry a list of mean and stdev for each task here too? You would need to change how the Scaler processed them, but that seems like the easier way to keep all this consistent. To keep things extra clear, you could create a new class of Scaler. Then StandardScaler and AtomBondScaler (or whatever the name was) could have a common Scaler class that they are subclasses of.

shihchengli replied:
I created AtomBondScaler class in scaler.py and normalize_atom_bond_targets() function in MoleculeDataset class. The  AtomBondScaler can carry a mean and a stdev for each task in atom- and bond-level targets. normalize_atom_bond_targets() is created to be distinguished from normalize_targets().

cjmcgill commented on chemprop/data/scaffold.py:
```diff
@@ -120,7 +120,7 @@ def scaffold_split(data: MoleculeDataset,
                      f'test scaffolds = {test_scaffold_count:,}')
 
-    if logger is not None:
+    if logger is not None and data[0].atom_targets is None and data[0].bond_targets is None:
```
What's this? Why is the logger disabled for atom/bond operation?

shihchengli replied:
I think atom- and bond-level targets don't need to process log_scaffold_stats funciton.

cjmcgill commented on chemprop/data/scaffold.py:
```diff
@@ -120,7 +120,7 @@ def scaffold_split(data: MoleculeDataset,
                      f'test scaffolds = {test_scaffold_count:,}')
 
-    if logger is not None:
+    if logger is not None and data[0].atom_targets is None and data[0].bond_targets is None:
```
Triggering logic off of `data[0].atom_targets is None` seems a bit complex for something you may want to check in lots of places. As long as we're not doing mixed molecule/atom targeting, it probably would make more sense to have `atom_bond_target: bool` as a dataset property.

cjmcgill commented on chemprop/features/featurization.py:
```diff
@@ -503,12 +507,14 @@ class BatchMolGraph:
     * :code:`max_num_bonds`: The maximum number of bonds neighboring an atom in this batch.
     * :code:`b2b`: (Optional) A mapping from a bond index to incoming bond indices.
     * :code:`a2a`: (Optional): A mapping from an atom index to neighboring atom indices.
+    * :code:`b2br`: (Optional): A mapping from f_bonds to real bonds in molecule recorded in targets.
```
What is a f_bond?

shihchengli replied:
A mapping from a bond index to a list of bond features.
https://github.com/chemprop/chemprop/blob/master/chemprop/features/featurization.py#L295

cjmcgill commented on README.md:
````diff
@@ -302,6 +303,20 @@ reaction and solvent/molecule encoding. Below are the input arguments for specif
+4      CC(=N)NCc1cccc(CNCc2ccncc2)c1  [-0.083162, 0.114954, -0.274544, -0.100369, 0....  ...  [[0.0, 1.5137126697008916, 2.4882198180715465,...  [[0.0, 1.0036, 0.0437, 0.0108, 0.0134, 0.0004,......
+```
+where atomic properties (e.g. hirshfeld_charges) must be a 1D list with the order same as that of atoms in the SMILES string; and bond properties (e.g. bond_length_matrix) can either be a 2D list of shape (number_of_atoms × number_of_atoms) or a 1D list with the order same as that of bonds in the SMILES string. If the input format is a pickle file, these targets can also be saved as numpy arrays.
+This model allows multitask constraints applied to different atomic/bond properties by specifying the argument `--constraints_path` with a given `.csv` file. Note that the constraints must be in the same order as the SMILES strings in your data file. Also note that `.csv` file must have a header row and the constraints should be comma-separated with one line per molecule. The `--adding_h` option  ...
````
Looks like I left a typo in my suggestion, sorry

```suggestion
This model allows multitask constraints applied to different atomic/bond properties by specifying the argument `--constraints_path` with a given `.csv` file. Note that the constraints must be in the same order as the SMILES strings in your data file. Also note that `.csv` file must have a header row and the constraints should be comma-separated with one line per molecule. The `--adding_h` option can be used if hydrogens are included in the atom targets and bonds to hydrogens are included in the bond targets.
```

shihchengli replied:
No problem! I missed this typo XD

cjmcgill commented on chemprop/data/utils.py:
```diff
@@ -298,19 +313,44 @@ def get_data(path: str,
 
+    # Load constraints
+    if constraints_path is not None:
+        constraints_data = []
+        reader = pd.read_csv(constraints_path)
+        reader_columns = reader.columns.tolist()
+        if len(reader_columns) != len(set(reader_columns)):
+            raise ValueError(f'There are duplicates in {constraints_path}.')
+        for target in args.target_columns:
+            if target in reader_columns:
+                constraints_data.append(reader[target].values)
+            else:
+                constraints_data.append([None] * len(reader))
+        constraints_data = np.transpose(constraints_data)  # each is num_data x num_targets
+
+        if args.save_smiles_splits:
+            raw_constraints_data = []
+            for target in reader_columns:
+                raw_constraints_data.append(reader[target].values)
+            raw_constraints_data = np.transpose(raw_constraints_data)  # each is num_data x num_columns
+        else:
+            raw_constraints_data = None
```
You don't have to, but may be worth considering breaking this out into a `get_constraints` function, like the `get_data_weights` function or similar.

shihchengli replied:
Thanks for this suggestion. By doing so, the code is more clear.

cjmcgill commented on chemprop/data/utils.py:
```diff
@@ -333,98 +373,166 @@ def get_data(path: str,
-    with open(path) as f:
+    extension = os.path.splitext(path)[1]
+    if extension == '.csv':
+        f = open(path)
```
`with open(path) as f:` is the preferred pattern because it doesn't require a subsequent `f.close()`. Not presently the case here, but if you had mulitple exits to a function, you would have to be careful to close f in each of them, and only once per exit even with complicated logic. The `with` version of the opening makes it easier to maintain going forward.

cjmcgill replied:
I understand that you changed it because the pkl reader is opening a file in a different way. And if we can't find another good way to do it, then that is a good enough reason to use `f=open(path)`. But I think worth the asking, is there a way to group these blocks so that you can use the `with` pattern?

cjmcgill replied:
It looks like `pd.read_pickle` can accept the buffer f as input so I think that simplifies all this and allows the `with` block right?

cjmcgill replied:
`pd.read_pickle` can take a binary-mode f object and `csv.DictReader` takes a text-mode f object. Trying to see if there's a way you could have them both use the same f. I think you could do something like

```
if extension in ['.pkl', '.pckl', '.pickle']:
    read_mode = 'rb'
else:  # extension = 'csv'
    read_mode = 'r'
with open(path, read_mode) as f:
...
```

mliu49 replied:
Is there a specific reason for adding support for pickle files? I think text-based formats are much more preferable in general.

Note that using `pd.read_pickle` also changes the general behavior here, such that the file will be loaded into memory in its entirely before being processed. This may be problematic for really large datasets.

cjmcgill replied:
@mliu49 the main reason would be to be consistent with atom and bond features which use pickled pandas to store the extra dimension for atom/bond. We want for the outputs of atom/bond predictions to be useable as features. We should probably add support for csv with lists in the atom/bond features inputs. Would we want to only support csv? We could say that csv is preferred for large datasets. 

In regards to the incremental loading, we might progressively load targets, but we load features all at once. And features, especially atom/bond features, are usually larger than the targets. With so many supplemental feature types now added, we might be due to refactor this function.

cjmcgill commented on chemprop/data/utils.py:
```diff
@@ -333,98 +373,166 @@ def get_data(path: str,
+                    for b in mol.GetBonds():
+                        bond_target_arranged.append(bond_target[b.GetBeginAtom().GetIdx(), b.GetEndAtom().GetIdx()])
+                    bond_targets.append(np.array(bond_target_arranged))
+            targets = atom_targets + bond_targets
```
Are there any situations where the ordering of targets would deviate from this in a way that makes problems? (I don't think so but worth asking the question).

shihchengli replied:
If the bond information provided by users is a 2D array whose ordering of atoms should match the RDKit molecule object. In the meantime, the ordering of bonds determined in L449-450 also matches the RDKit molecule object. That is consistent so I think it is fine.

cjmcgill commented on chemprop/data/utils.py:
```diff
@@ -333,98 +373,166 @@ def get_data(path: str,
+        reader = pd.read_pickle(path)
+        fieldnames = reader.columns.tolist()
+        reader = reader.iterrows()
+
```
After the `if csv` block and the `elif pkl` block, you should have a `else: raise NotImplementedError` block saying that the extension isn't a supported one.

cjmcgill commented on chemprop/data/utils.py:
```diff
@@ -668,8 +776,13 @@ def get_class_sizes(data: MoleculeDataset, proportion: bool = True) -> List[List
         for task_num in range(len(targets[i])):
-            if targets[i][task_num] is not None:
-                valid_targets[task_num].append(targets[i][task_num])
+            if data[0].atom_targets is not None or data[0].bond_targets is not None:
```
Good place for the dataset property that says if it's atom/bond or not

cjmcgill commented on chemprop/utils.py:
```diff
@@ -582,19 +602,28 @@ def save_smiles_splits(
     if not isinstance(smiles_columns, list):
         smiles_columns = preprocess_smiles_columns(path=data_path, smiles_columns=smiles_columns)
 
-    with open(data_path) as f:
```
If possible, I would recommend keeping the `with` block pattern like I suggested in data/utils

cjmcgill commented on chemprop/models/model.py:
```diff
@@ -33,7 +34,16 @@ def __init__(self, args: TrainArgs):
+        if self.is_atom_bond_targets:
+            self.atom_targets, self.bond_targets = args.atom_targets, args.bond_targets
+            self.atom_constraints, self.bond_constraints = args.atom_constraints, args.bond_constraints
+            self.device = args.device
```
Why is this separate device setting necessary?

shihchengli replied:
It would be used in L70 and L79 in train/predict.py.

cjmcgill replied:
Why doesn’t the model.to(device) in run_training or the device argument in load_checkpoint cover it like it does for molecule targets?

shihchengli replied:
`device` is not an attribute of nn.Module. A similar problem has been raised [here](https://github.com/pytorch/pytorch/issues/7460). As mentioned by Kani (in the comments), if all the parameters are on a single device, one could use `next(model.parameters()).device` to get the device of the model. I would try this way to remove the separate device setting.

cjmcgill commented on chemprop/models/model.py:
```diff
@@ -33,7 +34,16 @@ def __init__(self, args: TrainArgs):
+            self.atom_targets, self.bond_targets = args.atom_targets, args.bond_targets
+            self.atom_constraints, self.bond_constraints = args.atom_constraints, args.bond_constraints
+            self.device = args.device
+            self.output_size = 1
```
The `self.output_size` is a relative size multiplier for atom/bond targeting purposes and the base output size is set by the number of task names. This works, but now it means something different for the normal molecule targeting purposes. It would be better if the functions for these were made more parallel.

I think there are a few ways that you could make them parallel. The way that I would recommend would be:

- Make `self.output_size` relative for both of them. Set it initially as 1 for both cases. And maybe change the name to `self.relative_output_size` or something.
- Have the `output_size` input to DenseLayers be `args.num_tasks * self.relative_output_size`
- Change the inputs to MultiReadout to take numbers of tasks instead of the list of task names. That way we're not passing strings down further than we need to when we just need the number of how many, `len(args.atom_targets)` or similar.
- In the constraints loops, loop over the number of tasks rather than the task names.

cjmcgill commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,250 @@
+        :param atom_constraints: A list of booleans indicatin whether constraints applied to output of atomic properties.
+        :param bond_constraints: A list of booleans indicatin whether constraints applied to output of bond properties.
+        """
+        super(MultiReadout, self).__init__()
+        ind = 0
+        for i in range(len(atom_targets)):
+            constraint = atom_constraints[i]
+            self.add_module(f'readout_{ind}', FFNAtten(features_size, hidden_size, num_layers, output_size,
+                                                       dropout, activation, constraint, ffn_type='atom'))
+            ind += 1
+
+        for i in range(len(bond_targets)):
+            constraint = bond_constraints[i]
+            self.add_module(f'readout_{ind}', FFNAtten(2*features_size, hidden_size, num_layers, output_size,
+                                                       dropout, activation, constraint, ffn_type='bond'))
+            ind += 1
```
With this loop structure, do the different atom and bond tasks all have different FFN weights?

If so, then multitasking could have the different tasks cooperate on the MPN weights but all the FFN weights are independent?

mliu49 commented on chemprop/data/utils.py:
```diff
@@ -333,98 +373,166 @@ def get_data(path: str,
                     else:
                         raise ValueError('Inequality found in target data. To use inequality targets (> or <), the regression loss function bounded_mse must be used.')
+                elif '[' in value or ']' in value:
+                    target = np.array(eval(value))
```
It is generally a bad idea to use `eval` since it allows execution of arbitrary code. I'm not sure of the exact data structure expected here, but it may be possible to use `np.fromstring`.

shihchengli replied:
Thanks! This function is what I am looking for :)

davidegraff replied:
maybe `np.array(json.loads(value))`? `np.fromstring` doesn't expect `python` `list` syntax

shihchengli replied:
I tried to solve this problem by using regular expression but `np.array(json.loads(value))` is more elegant. Thanks for this suggestion!

cjmcgill commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,250 @@
+        elif self.ffn_type == 'bond':
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            hidden = torch.cat([forward_bond, backward_bond], dim=1)
```
This looks order dependent. I don't think we'd want the output of the bond prediction to depend on the arbitrary choice of which direction is considered forward. Chances are the model will converge for them to be similar, but we'd really like for them to be order invariant. Ideally, this would even help the model train faster because you wouldn't have to learn the weights for forward and reverse separately.

I'm not sure how we'd do that though. Maybe there's a way we can force the weights to be symmetrical across the two halves of the input. That would be the best way. We could also use some kind of aggregation function to add them together, but then we'd lose some information. I'll think on this and see if I can come up with anyting.

cjmcgill commented on chemprop/train/make_predictions.py:
```diff
@@ -293,11 +300,19 @@ def predict_and_save(
-            writer.writeheader()
-            for datapoint in full_data:
-                writer.writerow(datapoint.row)
+        if train_args.is_atom_bond_targets:
+            data = {}
+            for key in full_data[0].row.keys():
+                data[key] = [datapoint.row[key] for datapoint in full_data]
+            preds_dataframe = pd.DataFrame.from_dict(data)
+            preds_dataframe.to_pickle(args.preds_path)
```
Could we support saving the results both as .pkl or as csv? And maybe detect whether the extension provided is csv or pickle to choose between them?

cjmcgill commented on chemprop/train/run_training.py:
```diff
@@ -149,7 +154,12 @@ def run_training(args: TrainArgs,
-        scaler = train_data.normalize_targets()
+        if args.is_atom_bond_targets:
+            scaler = None
+            atom_bond_scalers = train_data.normalize_targets()
```
Oh I see, on the other side you're giving the list of scalers a different variable name. That could work.

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,389 @@
+        return output
+
+
+class DenseLayers(nn.Module):
```
this should not be a class- it's essentially a wrapper around an `nn.Sequential` object but it's hard for a reader to know that. Moreover, by reinventing the wheel (i.e., wrapping `nn.Sequential`) you create an opportunity to recreate the wheel incorrectly.

davidegraff replied:
you shoudl refactor this into a function `def ffn(d_input, ...) -> nn.Sequential`

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,389 @@
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
+        :param weights_ffn_num_layers: Number of layers in FFN for determining weights used to correct the constrained targets.
+        """
+        super(FFNAtten, self).__init__()
```
```suggestion
        super().__init__()
```

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,389 @@
+import re
+
+
+class AttrProxy(object):
```
```suggestion
class AttrProxy:
```

davidegraff replied:
is this class used anywhere else? If not, then it would be better served as just a helper method inside the `MultiReadout` class.

davidegraff replied:
It really looks like you're just hacking the `__getattr__` method of your `ffn` attribute below so its indexable like a `list`. You can just wrap the `ffn` inside a `ModuleList` object if you want that behavior

shihchengli replied:
Thanks for this valuable suggestion. Indeed, using `ModuleList` is more clear. I have made the change.

davidegraff reviewed (commented):
i need to spend more time going through the new model architecture code, but there's definitely a better way to write all this.

shihchengli commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
+        :param weights_ffn_num_layers: Number of layers in FFN for determining weights used to correct the constrained targets.
+        """
+        super(MultiReadout, self).__init__()
```
```suggestion
        super().__init__()
```

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+                if size == 0:
+                    continue
+                else:
+                    cur_weights = weights.narrow(0, start, size)
+                    cur_output = output.narrow(0, start, size)
```
use tensor indexing here rather than `narrow`

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+                    cur_weights = weights.narrow(0, start, size)
+                    cur_output = output.narrow(0, start, size)
+
+                    cur_weights = torch.nn.Softmax(dim=0)(cur_weights).reshape(-1,)
```
```suggestion
                    cur_weights = F.softmax(cur_weights, dim=0).flatten()
```

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
```
what happens if `self.ffn_type` is neither?

shihchengli replied:
`ffn_type` can only be `atom` or `bond`. The `FFNAtten` class will be constructed in the initialization of `MultiReadout()` class so the ffn_type is already given inherently. However, some users might want to use `FFNAtten` class directly for some purposes, so I will add `ValueError` here if other values are specified.

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+    activation: nn.Module,
+    dataset_type: str = None,
+    spectra_activation: str = None,
+):
```
add return type

shihchengli replied:
Added

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+        spectra_activation = nn.Softplus() if spectra_activation == "softplus" else nn_exp()
+        layers.append(spectra_activation)
+
+    ffn = nn.Sequential(*layers)
+    return ffn
```
```suggestion
    return nn.Sequential(*layers)
```

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+        :param bond_types_batch: A list of PyTorch tensors storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`MultiReadout`, a list of PyTorch tensors which ontains atomic/bond properties prediction.
+        """
+        return [ffn(input, constraints_batch[i], bond_types_batch[i]) for i, ffn in enumerate(self.ffn_list)]
```
can this be a tensor return type?

shihchengli replied:
Both atomic and bond properties can be trained simultaneously, and the length of the returned vector for an atomic property would be different than that for a bond property. So I think it cannot be a tensor type here. It is still possible to get a padded tensor, but it would be too complicated for the following processing (need to additionally define a mask and pass it for the following processes).

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+
+class MultiReadout(nn.Module):
+    """A fake list of FFNs for reading out as suggested in
+    https://discuss.pytorch.org/t/list-of-nn-module-in-a-nn-module/219/3"""
```
this link is broken

shihchengli replied:
This should be removed when I replace `AttrProxy()` with `nn.ModuleList()`. I will remove it.

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+            self.atom_ffn_base = nn.Identity()
+            self.bond_ffn_base = nn.Identity()
+
+        ind = 0
```
what's this variable doing?

shihchengli replied:
Oops! This is also a legacy of previously removed `AttrProxy()`. Thanks for noticing this.

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+
+    def forward(
+        self,
+        input: Tuple[torch.tensor, List, torch.tensor, List, torch.tensor],
```
`torch.tensor` is a function not a class

shihchengli replied:
fixed

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,355 @@
+
+                    cur_weights = torch.nn.Softmax(dim=0)(cur_weights).reshape(-1,)
+
+                    cur_output_sum = cur_output[:, 0].sum()
+
+                    cur_output[:, 0] = cur_output[:, 0] + cur_weights * (constraints[i] - cur_output_sum)
```
```suggestion
                    cur_output[:, 0] = cur_output[:, 0] + cur_weights * (constraints[i] - cur_output[:, 0].sum())
```
it's more readable as a one-liner

davidegraff reviewed (commented):
slowly working my way through

davidegraff commented:
more broadly, can you make the notation in the code closer to the paper in `ffn.py`- it's hard to compare the code to the math of the original paper when the symbols don't quite match up

shihchengli commented:
I modified some notations in the code. Hopefully, it is more clear now. Actually, Chas and I added some flexibility in this model compared to the original one from the paper. To mimic the one in the paper, the attribute of `shared_ffn` is False and `weights_ffn_num_layers` is 2. In this case, the calculated weights for constrained learning were simply obtained via 2-layer FFNN with corresponding atomic/bond feature vector as input, and the outputs were further passed to a Softmax function to get the weights. The 2-layer FFNN will return the value of $u\hat{\mathbf{a}_{i}}$ in Equation 14 in SI.

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,350 @@
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+        else:
+            raise ValueError(f"Unrecognized ffn_type of {self.ffn_type}.")
```
```suggestion
            raise RuntimeError(f"Unrecognized ffn_type of {self.ffn_type}.")
```

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,350 @@
+        return output
+
+
+class nn_exp(nn.Module):
```
```suggestion
class Exp(nn.Module):
```
classes should be CamelCase

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,350 @@
+                if bond_types is not None:
+                    output = output + bond_types.reshape(-1, 1)
+
+            u_a = self.weights_readout(output_hidden)
```
```suggestion
            W_a = self.weights_readout(output_hidden)
```
the formula in the paper uses `W`

shihchengli replied:
This is the $u\hat{a_i}$ in equation 14 in the SI. It would become ${w_i}$ after being passed in the softmax function.

davidegraff replied:
why even bother with it then? `u_a` here is only used in the context of being passed into the `Softmax` function and the term only appears in the SI. If we can avoid terms from the SI in favor of only using terms from the main text (as we can here), we should do so

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,350 @@
+                if size == 0:
+                    continue
+                else:
+                    u_ai = u_a[start:start+size]
```
```suggestion
                    w_ai = W_a[start:start+size]
```

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,350 @@
+                    qi = output[start:start+size]
+                    Q = constraints[i]
+
+                    wi = F.softmax(u_ai, dim=0).flatten()
+
+                    qi[:, 0] = qi[:, 0] + wi * (Q - qi[:, 0].sum())
+                    constrained_output.append(qi)
```
do we need to do all this indexing and slicing and flattening? seems messy and i think broadcasting will work just fine here

shihchengli replied:
I think the slicing and indexing are necessary because the softmax will act on the $u\hat{a_i}$  of atoms in a molecule. The purpose of flattening the output of softmax is to let the shape of `wi` be the same as  `qi[:, 0]`. What do you think about using broadcasting here?

shihchengli replied:
`wi`: (number_of_atoms)
`qi`: (number_of_atoms, 1)
`u_ai `: (number_of_atoms, 1)
`output`: (sum_of_number_of_atoms, 1)

PS. The `sum_of_number_of_atoms` is the sum of number of atoms for each molecule in this batch.

davidegraff replied:
```python
>>> n_atoms = 10
>>> Q = 1
>>> q_i = torch.rand(n_atoms, 1)
>>> u_ai = torch.rand(n_atoms, 1)
>>> w_i = u_ai.softmax(0)
# your code
>>> # separating the assignment into two steps
>>> q_f1 = q_i[:, 0] + w_i.flatten() * (Q - q_i[:, 0].sum())
>>> q_f1.shape
torch.Size([10])
>>> # assignment would happen here: q_i[:, 0] = q_f1
# suggestion
>>> q_f = q_i + w_i * (Q - q_i.sum())
>>> q_f.shape
torch.Size([10, 1])
>>> q_f2[:, 0].equal(q_f1)
True
>>> constrained_output.append(q_f2)

davidegraff reviewed (commented):
a broader question here on the object model- why is `constraint` a parameter to the `FFNAtten` class? It seems like the class will break if `constraint` is `False` but then a user passes in input with non-null `constraints`.

shihchengli commented:
For different atom/bond targets, each of them would have its own FFN model. Some of them might need constraints on the outputs, and some of them might not. For properties that need constraints, each of them would have an additional attention-based layer. The `constraint` attribute can let the `FFNAtten` class know how to initialize the model architecture. You are totally right that it will break if constraint is False but then a user passes in input with non-null constraints. I think it would raise an error like `self.weights_readout is undefined`. I just come up with an idea to avoid this unintended behavior happening. I should replace the statement of `if constraints is not None:` in L251 with `if self.constraint:` (See the code suggestion below). Does it make sense?

shihchengli commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,350 @@
+        else:
+            raise ValueError(f"Unrecognized ffn_type of {self.ffn_type}.")
+
+        if constraints is not None:
```
```suggestion
       if self.constraint:
```

davidegraff replied:
I think the broader question is the level of abstraction here. Your object model makes it such that the atom/bond-level prediction always passes through an `FFNAtten` module _regardless of whether it is constrained_. Logically, the `FFNAtten` module exists for the purposes of **constrained** output, so why are using it to make _unconstrained_ predictions?

shihchengli replied:
Added!

shihchengli commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,350 @@
+
+    # If spectra model, also include spectra activation
+    if dataset_type == "spectra":
+        spectra_activation = nn.Softplus() if spectra_activation == "softplus" else nn_exp()
```
```suggestion
        spectra_activation = nn.Softplus() if spectra_activation == "softplus" else Exp()
```

shihchengli commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,350 @@
+                    qi = output[start:start+size]
+                    Q = constraints[i]
+
+                    wi = F.softmax(u_ai, dim=0).flatten()
```
```suggestion
                    wi = F.softmax(w_ai, dim=0).flatten()
```

davidegraff reviewed (commented):
why is the `dropout` argument generally of type `nn.Module`? It seems unnecessary to compose these objects using a `Dropout` module rather than just the dropout probability

shihchengli commented:
You are totally right. The change will be made soon. Thank you for reviewing this part so carefully. It looks more logical and clear now.

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,439 @@
+                ),
+                activation,
+            )
+        else:
+            self.atom_ffn_base = nn.Identity()
+            self.bond_ffn_base = nn.Identity()
```
why assign these to the identity function if they won't be used? if `shared_ffn` is `False`, it's passed to all initializers and then these objects are ignored

shihchengli replied:
Yes... I will replace those with `None`

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,439 @@
+        else:
+            base_output_size = hidden_size
+
+        if shared_ffn:
+            self.ffn = ffn_base
```
can we condense these two a single argument of type `Optional[Module]`. I.e., **if** an `ffn_base` is passed in, this class will use it

shihchengli replied:
This is a good suggestion. I will apply it. Thank you!

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,439 @@
+                )
+            else:
+                self.ffn = nn.Identity()
+        self.ffn_readout = build_ffn(
+            first_linear_dim=base_output_size,
+            hidden_size=hidden_size,
+            num_layers=1,
+            output_size=output_size,
+            dropout=dropout,
+            activation=activation,
+        )
```
this is just a linear layer, why build it using some abstract function?

shihchengli replied:
The reason I wrote it is for consistency (building every FFN via `build_ffn()` function), but this one is relatively simpler so I should replace this block with the below code.
```
self.ffn_readout = nn.Sequential(
    nn.Dropout(dropout),
    nn.Linear(base_output_size, output_size),
)
```

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,439 @@
+    num_layers: int,
+    output_size: int,
+    dropout: float,
+    activation: nn.Module,
```
can we use the `get_activation()` function defined in this package? Seems unnatural that some functions in the library would have the `activation` argument correspond to the _name_ of the activation function whereas for some functions it actually corresponds to the `Module` of the activation function

shihchengli replied:
That makes sense to me. I will apply this suggestion. Thanks a lot!

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,439 @@
+        """
+        super().__init__()
+
+        if num_layers == 1:
+            base_output_size = features_size
+        else:
+            base_output_size = hidden_size
```
```suggestion
        base_output_size = features_size if num_layers == 1 else hidden_size
```

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,439 @@
+
+        return output
+
+class FFNAtten(nn.Module):
```
this class can be defined as a subclass of `FFN` to avoid code duplication

shihchengli replied:
I think those two classes are quite different either in __init__() and forward(). There is not too much code duplication in those two classes.

davidegraff replied:
It's only "different" because you've organized them be that way. `FFNAtten` is a logical extension of `FFN` if you reorganize it correctly:
```python
class FFN(nn.Module):
    def __init__(
        self,
        features_size: int,
        hidden_size: int,
        num_layers: int,
        output_size: int,
        dropout: float,
        activation: str,
        ffn_base: Optional[nn.Module] = None,
        ffn_type: str = "atom",
    ):
        super().__init__()

        base_output_size = features_size if num_layers == 1 else hidden_size

        if ffn_base:
            self.ffn = ffn_base
        else:
            if num_layers > 1:
                self.ffn = nn.Sequential(
                    build_ffn(
                        first_linear_dim=features_size,
                        hidden_size=hidden_size,
                        num_layers=num_layers - 1,
                        output_size=hidden_size,
                        dropout=dropout,
                        activation=activation,
                    ),
                    get_activation_function(activation),
                )
            else:
                self.ffn = nn.Identity()
        
        self.ffn_readout = nn.Sequential(
            nn.Dropout(dropout),
            nn.Linear(base_output_size, output_size),
        )
        self.ffn_type = ffn_type
```

which can now be seen to be the logical superclass of `FFNAtten`:
```python
class FFNAtten(nn.Module):
    def __init__(
        self,
        features_size: int,
        hidden_size: int,
        num_layers: int,
        output_size: int,
        dropout: float,
        activation: str,
        ffn_base: Optional[nn.Module] = None,
        ffn_type: str = "atom",
        weights_ffn_num_layers: int = 2,
    ):
        super().__init__()

        base_output_size = features_size if num_layers == 1 else hidden_size

        if ffn_base:
            self.ffn = ffn_base
        else:
            if num_layers > 1:
                self.ffn = nn.Sequential(
                    build_ffn(
                        first_linear_dim=features_size,
                        hidden_size=hidden_size,
                        num_layers=num_layers - 1,
                        output_size=hidden_size,
                        dropout=dropout,
                        activation=activation,
                    ),
                    get_activation_function(activation),
                )
            else:
                self.ffn = nn.Identity()
        # build a linear readout from the hidden outputs
        self.ffn_readout = nn.Sequential(
            nn.Dropout(dropout),
            nn.Linear(base_output_size, output_size),
        )
        self.ffn_type = ffn_type
        self.weights_readout = build_ffn(
            first_linear_dim=base_output_size,
            hidden_size=hidden_size,
            output_size=1,
            num_layers=weights_ffn_num_layers,
            dropout=dropout,
            activation=activation,
        )
```
and rewritten using inheritance:
```python
class FFNAtten(FFN):
    def __init__(
        self,
        features_size: int,
        hidden_size: int,
        num_layers: int,
        output_size: int,
        dropout: float,
        activation: str,
        ffn_base: Optional[nn.Module] = None,
        ffn_type: str = "atom",
        weights_ffn_num_layers: int = 2,
    ):
       super().__init__(
            features_size,
            hidden_size,
            num_layers,
            output_size,
            dropout,
            activation,
            ffn_base,
            ffn_type
        )
        self.weights_readout = build_ffn(
            first_linear_dim=base_output_size,
            hidden_size=hidden_size,
            output_size=1,
            num_layers=weights_ffn_num_layers,
            dropout=dropout,
            activation=activation,
        )
```
the same can be done for `forward()` (with a bit more work). The reason this matters is inheritance makes it easier to reason about logical relationships between classes/methods. You can see this even in the figure you've created to explain atom/bond-level target prediction. We use the _same_ base FFN for both constrained and unconstrained predictions, with some seasoning **on top** for constrained predictions. Therefore, we should write the code to reflect how we reason about the concept.

davidegraff replied:
you can also refactor `FFN.forward()` to be comprised of two calls: (1) `calc_hidden()`, which returns the initial hidden outputs and (2) `readout`, which takes the hidden outputs to a final output. `FFNAtten` would then only need to override `readout` because both networks calculate the initial hiddens in the same and only differ in how they take these hiddens to a final output. Again, this reflects how we _present_ the concept to readers in both text and figures.

shihchengli replied:
Thanks for guiding me through this. I made the change now.

davidegraff commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,439 @@
+            if constraint:
+                ffn_list.append(
+                    FFNAtten(
+                        features_size=atom_features_size,
+                        hidden_size=atom_hidden_size,
+                        num_layers=num_layers,
+                        output_size=output_size,
+                        dropout=dropout,
+                        activation=activation,
+                        ffn_base=self.atom_ffn_base,
+                        ffn_type="atom",
+                        shared_ffn=shared_ffn,
+                        weights_ffn_num_layers=weights_ffn_num_layers,
+                    )
+                )
+            else:
+                ffn_list.append(
+                    FFN(
+                        features_size=atom_features_size,
+                        hidden_size=atom_hidden_size,
+                        num_layers=num_layers,
+                        output_size=output_size,
+                        dropout=dropout,
+                        activation=activation,
+                        ffn_base=self.atom_ffn_base,
+                        ffn_type="atom",
+                        shared_ffn=shared_ffn,
```
both of these objects are parameterized very similarly. Can you pull the common arguments out into some parameter dict `ffn_params` and then pass it in using dictionary unpacking (`**ffn_params`)

shihchengli replied:
Sounds good! I should do that.

davidegraff reviewed (commented):
getting closer on a more streamlined object model

shihchengli commented on chemprop/models/ffn.py:
```diff
@@ -0,0 +1,439 @@
+        """
+        super().__init__()
+
+        if num_layers == 1:
+            base_output_size = features_size
+        else:
+            base_output_size = hidden_size
```
```suggestion
        base_output_size = features_size if num_layers == 1 else hidden_size
```

davidegraff commented on chemprop/data/utils.py:
```diff
@@ -50,8 +55,8 @@ def preprocess_smiles_columns(path: str,
     else:
-        if not isinstance(smiles_columns,list):
-            smiles_columns=[smiles_columns]
+        if not isinstance(smiles_columns, list):
+            smiles_columns = [smiles_columns]
```
what about the case where a `Sequence[str]` is passed? that fails this check and then is assumed to be of type `str`. Type-checking at runtime is very tenuous in python, so it's best practice to check on the most _restrictive_ type, which in this case is: `if isinstance(smiles_columns,str)`

davidegraff reviewed (commented):
making my way through.

shihchengli commented on chemprop/models/model.py:
```diff
@@ -189,32 +206,59 @@ def forward(self,
+                lambdas = self.softplus(lambdas)  # + min_val
+                alphas = self.softplus(alphas) + 1  # + min_val # add 1 for numerical contraints of Gamma function
+                betas = self.softplus(betas)  # + min_val
+                return torch.cat([means, lambdas, alphas, betas], dim=1)
```
You are right! I left what it was before.

shihchengli commented on chemprop/models/mpn.py:
```diff
@@ -59,23 +60,34 @@ def __init__(self, args: TrainArgs, atom_fdim: int, bond_fdim: int, hidden_size:
                 mol_graph: BatchMolGraph,
-                atom_descriptors_batch: List[np.ndarray] = None) -> torch.FloatTensor:
+                atom_descriptors_batch: List[np.ndarray] = None,
+                bond_descriptors_batch: List[np.ndarray] = None) -> torch.FloatTensor:
```
Fixed!

shihchengli commented on chemprop/models/mpn.py:
```diff
@@ -37,6 +37,7 @@ def __init__(self, args: TrainArgs, atom_fdim: int, bond_fdim: int, hidden_size:
+        self.is_atom_bond_targets = args.is_atom_bond_targets
 
         # Dropout
         self.dropout_layer = nn.Dropout(p=self.dropout)
```
I totally agree with you. The change is made.

shihchengli commented on chemprop/args.py:
```diff
@@ -181,6 +192,15 @@ def bond_features_size(self) -> int:
+
+    @bond_descriptors_size.setter
+    def bond_descriptors_size(self, bond_descriptors_size: int) -> None:
+        self._bond_descriptors_size = bond_descriptors_size
```
`_bond_descriptors_size` is a private variable defined at line 132.

davidegraff replied:
yes but `_bond_descriptors_size` is not _actually_ a private attribute. A client can easily access/modify it just like a normal attribute. You need a double underscore ("dunder", `__`) to actually make it "private" (quotes because nothing is ever truly private in python if you try hard enough)

davidegraff commented on chemprop/features/featurization.py:
```diff
@@ -556,17 +556,17 @@ def __init__(self, mol_graphs: List[MolGraph]):
-        self.a2b = torch.LongTensor([a2b[a] + [0] * (self.max_num_bonds - len(a2b[a])) for a in range(self.n_atoms)])
-        self.b2a = torch.LongTensor(b2a)
-        self.b2revb = torch.LongTensor(b2revb)
+        self.f_atoms = torch.tensor(f_atoms, dtype=torch.float32)
+        self.f_bonds = torch.tensor(f_bonds, dtype=torch.float32)
```
```suggestion
        self.f_atoms = torch.tensor(f_atoms, dtype=torch.float)
        self.f_bonds = torch.tensor(f_bonds, dtype=torch.float)
```

davidegraff reviewed (commented):
new code is looking better but it looks like it's failing some tests now? Can you check on your local machine if something hasn't broken?

davidegraff commented on chemprop/models/model.py:
```diff
@@ -93,62 +101,65 @@ def create_ffn(self, args: TrainArgs) -> None:
-
-        # Create FFN model
-        self.ffn = nn.Sequential(*ffn)
+            self.readout = build_ffn(first_linear_dim=atom_first_linear_dim,
```
was this change in the model submodule name intentional? I don't disagree with the naming (I think it's more logically correct), but it will *break* module saving/loading for older versions of chemprop models because torch internals load modules based on key names

shihchengli replied:
The backward compatibility for parameter names is taken care in chemprop/utils.py.

davidegraff replied:
that seems like overkill to change a module name when the change isn't strictly necessary in the first place. Also, something that would need to be discussed with the dev team at large.

shihchengli replied:
I totally agree. How about still sticking with the original name here?

davidegraff commented on chemprop/models/mpn.py:
```diff
@@ -59,23 +59,34 @@ def __init__(self, args: TrainArgs, atom_fdim: int, bond_fdim: int, hidden_size:
         # Shared weight matrix across depths (default)
         self.W_h = nn.Linear(w_h_input_size, self.hidden_size, bias=self.bias)
 
+        # hidden state readout
```
delete comments

davidegraff commented on chemprop/models/mpn.py:
```diff
@@ -59,23 +59,34 @@ def __init__(self, args: TrainArgs, atom_fdim: int, bond_fdim: int, hidden_size:
+        if self.is_atom_bond_targets:
+            self.W_o_b = nn.Linear(self.bond_fdim + self.hidden_size, self.hidden_size)
+
         # layer after concatenating the descriptors if args.atom_descriptors == descriptors
```
delete

davidegraff commented on chemprop/models/mpn.py:
```diff
@@ -59,23 +59,34 @@ def __init__(self, args: TrainArgs, atom_fdim: int, bond_fdim: int, hidden_size:
             self.atom_descriptors_size = args.atom_descriptors_size
             self.atom_descriptors_layer = nn.Linear(self.hidden_size + self.atom_descriptors_size,
                                                     self.hidden_size + self.atom_descriptors_size,)
+        # layer after concatenating the descriptors if args.bond_descriptors == descriptors
```
delete

davidegraff commented on chemprop/models/mpn.py:
```diff
@@ -170,7 +213,7 @@ def __init__(self,
         self.reaction = args.reaction
         self.reaction_solvent = args.reaction_solvent
         self.atom_fdim = atom_fdim or get_atom_fdim(overwrite_default_atom=args.overwrite_default_atom_features,
-                                                     is_reaction=(self.reaction or self.reaction_solvent))
+                                                    is_reaction=(self.reaction or self.reaction_solvent))
         self.bond_fdim = bond_fdim or get_bond_fdim(overwrite_default_atom=args.overwrite_default_atom_features,
                                                     overwrite_default_bond=args.overwrite_default_bond_features,
                                                     atom_messages=args.atom_messages,
```
using logical `or` with primitive data is tenuous:
```python
>>> None or 2
2
>>> 0 or 2
2
```
I'd argue the intention with the second case was to get `0` out. It's best to be explicit here.

shihchengli replied:
Those two parameters are with boolean values. I replace the original line with `is_reaction=self.reaction if self.reaction is not False else self.reaction_solvent`.

davidegraff commented on chemprop/models/mpn.py:
```diff
@@ -265,22 +311,22 @@ def forward(self,
+                    else:
+                        encodings.append(self.encoder_solvent(ba))
 
         output = reduce(lambda x, y: torch.cat((x, y), dim=1), encodings)
```
I know this line isn't your doing, but isn't this just doing the following:
```python
xs: Iterable[Tensor] = [x1, x2, x3]
X = torch.cat((torch.cat((x1, x2), dim=1), x3), dim=1)
```
I.e., progressively concatenating the next tensor onto the result of concatenating the previous two tensors? If so, why not just call `torch.cat(encodings, dim=1)`

shihchengli replied:
I thought that is progressively concatenating. I have modified the code.

davidegraff reviewed (commented):
can you format the new model files with `black` since we adopted that code format?

shihchengli commented:
For atom/bond level properties prediction, using atom-mapped SMILES is intuitive so I will additionally add some code in the PR.

oscarwumit commented:
@shihchengli Can you resolve the conflicts and rebase?

shihchengli commented:
@oscarwumit FIxed!

oscarwumit commented:
@shihchengli Thanks. Can you also squash similar commits into one?

oscarwumit commented on chemprop/multitask_utils.py:
```diff
@@ -0,0 +1,84 @@
+    """
+    n_atoms, n_bonds = test_data.number_of_atoms, test_data.number_of_bonds
+    individual_values = [[[] for j in range(num_tasks)] for i in range(len(test_data))]
+    for i in range(natom_targets):
```
Nested for loops are known to be inefficient. Since `natom_targets`, `nbond_targets`, and `num_models` could potentially be large, the loops here might take a while.

Maybe consider refactoring using numpy built-in functionalities to achieve the desired reshape. If it takes too much time, we can keep this for now. Thanks.

shihchengli replied:
That is the thing I should do. I will try to refactor it.

shihchengli replied:
@oscarwumit I have refactored those functions. The code should be more efficient now.

oscarwumit replied:
thanks

oscarwumit reviewed (commented):
Thanks for the PR. I left some minor comments. Can you address them before we do a final check and merge?

shihchengli commented on chemprop/data/data.py:
```diff
@@ -67,42 +69,55 @@ def __init__(self,
+        :param atom_descriptors: A numpy array containing additional atom descriptors to featurize the molecule.
+        :param bond_descriptors: A numpy array containing additional bond descriptors to featurize the molecule.
+        :param constraints: A numpy array containing constraints applied to different atomic/bond properties.
+        :param raw_constraints: A numpy array containing constraints in raw data.
```
`raw_constraints` is a numpy array that contains every constraint presented in row data, even if the corresponding columns are not used as targets. On the other hand, `constraints` contains only a subset of raw_constraints corresponding to the targets. The reason to preserve `raw_constraints` is because the raw constraints for each split can be saved when `--save_smiles_splits` flag is used.

oscarwumit replied:
Thanks. For clarity, maybe consider revising the parameter description to:

`
:param raw_constraints: A numpy array containing all user-provided atom/bond-level constraints in input data.
`
`
:param constraints: A numpy array containing atom/bond-level constraints that are used in training. Param constraints is a subset of param raw_constraints.
`

shihchengli commented on chemprop/args.py:
```diff
@@ -103,7 +103,13 @@ class CommonArgs(Tap):
+    """
+    Custom extra bond descriptors.
+    :code:`feature`: used as bond features to featurize a given molecule.
+    :code:`descriptor`: used as descriptor and concatenated to the machine learned atomic representation.
```
Yeah! That is a typo.

shihchengli commented on chemprop/args.py:
```diff
@@ -393,6 +421,28 @@ class TrainArgs(CommonArgs):
+    Whether RDKit molecules keep the original atom mapping. This option is intended to be used when providing atom-mapped SMILES with
+    the :code:`is_atom_bond_targets`.
+    """
+    is_atom_bond_targets: bool = False
```
That does make sense. The change has applied.

shihchengli commented on chemprop/args.py:
```diff
@@ -548,14 +639,42 @@ def process_args(self) -> None:
+                keep_atom_map=self.keeping_atom_map,
+            )
+            self.target_columns = self.atom_targets + self.bond_targets
+            # self.target_columns = self.atom_targets + self.bond_targets + self.molecule_targets  # TODO: Support mixed targets
```
Thanks for the good suggestion. The README.md has been updated.

oscarwumit reviewed (approved):
Thanks for the PR.
